src/Storages/MergeTree/calculateBlockNumbersForRestoring.cpp

```diff
@@ -0,0 +1,482 @@
+
+        /// Prepare list of permutations so we can apply it two vectors `mutations` and `mutation_names_in_backup`.
+        std::vector<size_t> permutation;
+        permutation.resize(count);
+        std::iota(permutation.begin(), permutation.end(), 0);
+
+        std::sort(
+            permutation.begin(),
+            permutation.end(),
+            [&](size_t left_index, size_t right_index)
+            {
+                const auto & left = mutations[left_index];
+                const auto & right = mutations[right_index];
+                return left.number < right.number;
+            });
+
+        apply_permutation(mutations, mutation_names_in_backup, std::move(permutation));
+    }
+
+    struct PartMutations
+    {
+        /// Index of the first mutation this part is applied to or should be applied to.
+        ///
+        /// For example, let's consider a MergeTree with six parts "all_1_2_1_8", "all_3_3_0", "all_4_4_0_5", "all_6_6_0_8", "all_7_7_0", "all_9_9_0"
+        /// and two mutations "mutation_5.txt", "mutation_8.txt". For those parts `first_mutation` and `num_applied_mutations` will be calculated as:
+        /// | part_name   | first_mutation | num_applied_mutations |
+        /// | all_1_2_1_8 | 0 (mutation_5) |                     2 |
+        /// | all_3_3_0   | 0              |                     0 |
+        /// | all_4_4_0_5 | 0              |                     1 |
+        /// | all_6_6_0_8 | 1 (mutation_8) |                     1 |
+        /// | all_7_7_0   | 1              |                     0 |
+        /// | all_9_9_0   | 2 (none)       |                     0 |
+        ///
+        size_t first_mutation = 0;
+
+        /// Number of mutations this part is already applied to.
+        size_t num_applied_mutations = 0;
+    };
+
+    /// Find corresponding mutations for each part.
+    std::vector<PartMutations>
+    findMutationsForEachPart(const std::vector<MergeTreePartInfo> & parts, const std::vector<MutationInfoFromBackup> & mutations)
+    {
+        std::vector<PartMutations> res;
+        res.resize(parts.size());
+
+        if (mutations.empty())
+        {
+            /// No mutations.
+        }
+        else if (mutations.front().block_number)
+        {
+            /// Mutations are specified with a single `block_number` per a mutation.
+
+            /// Make a map of mutations where mutations are ordered by their block numbers.
+            /// This map needs to be ordered because we use upper_bound() below.
+            std::map<Int64, size_t> mutations_by_block_number;
+
+            for (size_t i = 0; i != mutations.size(); ++i)
+            {
+                const auto & mutation = mutations[i];
+
+                /// MergeTree and ReplicatedMergeTree mutations can't be mixed.
+                if (!mutation.block_number || mutation.block_numbers)
+                    throw Exception(ErrorCodes::CANNOT_RESTORE_TABLE, "Mutations specify incorrect block numbers");
```

Comment:
This situation is possible because a backup can be damaged or written in a wrong way. I've changed the error message to make it more clear, see https://github.com/ClickHouse/ClickHouse/pull/50230/commits/d701dad2e70800f64c06361f8dcced8a78f5f5b1#diff-f56a47794b5c9778b9c556395f387a619390917a467a67be341242dda79b5a77R63

Reply:
It still makes sense to use `LOGICAL_ERROR` code in such cases with an error message like https://github.com/ClickHouse/ClickHouse/blob/3a1182f6c02fd4f41bc7f66c8cd5ee9182e91e22/src/Storages/MergeTree/ActiveDataPartSet.cpp#L81-L82
e.g. "Mutations specify incorrect block numbers. Either backup data is corrupted or it's a bug"

Because usually backups should not be corrupted unless we have a bug

Reply:
Indeed, usually they should not be corrupted. Usually but not always. Backups can be copied to another location or moved around. Maybe it's better to invent another error code - something like `ErrorCodes::BACKUP_CORRUPTION`